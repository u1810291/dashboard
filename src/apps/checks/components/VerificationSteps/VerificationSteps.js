import { Box, Button, IconButton, Typography } from '@material-ui/core';
import { useConfirm, useOverlay } from 'apps/overlay';
import { useSelector } from 'react-redux';
import { difference, without } from 'lodash';
import { getDocumentList } from 'models/Document.model';
import React, { useCallback, useState } from 'react';
import { selectDenialUploadAvailability } from 'state/merchant/merchant.selectors';
import { FiEdit2, FiTrash2 } from 'react-icons/fi';
import { useIntl } from 'react-intl';
import { VerificationStepsModal } from '../VerificationStepsModal/VerificationStepsModal';
import { useStyles } from './VerificationSteps.styles';
import { DenyUploadRequirement } from '../../../configuration/containers/DenyUploadRequirement/DenyUploadRequirement';

export function removeItem(steps, index) {
  const updatedSteps = [...steps];
  updatedSteps.splice(index, 1);
  return updatedSteps;
}

export function replaceItem(steps, index, values) {
  const updatedSteps = [...steps];
  updatedSteps[index] = values;
  return updatedSteps;
}

export function addItem(steps, values) {
  const updatedSteps = [...steps];
  updatedSteps.push(values);
  return updatedSteps;
}

export function accessibleItems(available, steps, index) {
  return difference(available, ...without(steps, steps[index]));
}

export function VerificationSteps({ steps = [], onChange }) {
  const intl = useIntl();
  const classes = useStyles();
  const [createOverlay, closeOverlay] = useOverlay();
  const confirm = useConfirm();
  const [availableDocumentTypes] = useState(getDocumentList());
  const isUploadDenyAvailable = useSelector(selectDenialUploadAvailability);

  const handleRemoveItem = useCallback(async (index) => {
    try {
      await confirm(intl.formatMessage({ id: 'confirm_string' }));
      onChange({ verificationSteps: removeItem(steps, index) });
    } catch (e) {
      // none, canceled
    }
  }, [confirm, intl, onChange, steps]);

  const onEditItem = useCallback((index) => {
    createOverlay(
      <VerificationStepsModal
        values={index !== undefined ? steps[index] : []}
        items={accessibleItems(
          availableDocumentTypes,
          steps,
          index,
        )}
        onSave={(values) => {
          closeOverlay();
          onChange({
            verificationSteps:
              index !== undefined
                ? replaceItem(steps, index, values)
                : addItem(steps, values),
          });
        }}
      />,
    );
  }, [createOverlay, onChange, availableDocumentTypes, closeOverlay, steps]);

  return (
    <Box>
      <Typography variant="h4" gutterBottom>
        {intl.formatMessage({ id: 'flow.documentTypeStep.title' })}
      </Typography>

      {steps.map((step, index) => (
        <Box key={index} mt={2}>
          <Typography variant="h6" color="primary" gutterBottom>
            {intl.formatMessage({ id: 'flow.documentTypeStep.stepNo' })}
            {' - '}
            {index + 1}
          </Typography>

          <Box display="flex" mt={1}>
            {/* step list */}
            <Box flexGrow={1}>
              {step.sort().map((item) => (
                <Box key={item} className={classes.value}>
                  <Typography variant="body1" gutterBottom>
                    {intl.formatMessage({ id: `flow.documentTypeStep.${item}` })}
                  </Typography>
                </Box>
              ))}
            </Box>
            {/* actions */}
            <Box flexGrow={0} display="flex" ml={1} alignItems="center">
              <IconButton size="small" color="primary" onClick={() => onEditItem(index)}>
                <FiEdit2 />
              </IconButton>
              <Box ml={1}>
                <IconButton size="small" onClick={() => handleRemoveItem(index)}>
                  <FiTrash2 className="color-red" />
                </IconButton>
              </Box>
            </Box>
          </Box>
        </Box>
      ))}

      {difference(availableDocumentTypes, ...steps).length > 0 && (
        <Box mt={2}>
          <Button
            variant="text"
            color="primary"
            onClick={() => onEditItem()}
            disabled={difference(availableDocumentTypes, ...steps).length === 0}
          >
            {intl.formatMessage({ id: 'flow.documentTypeStep.button.title' })}
          </Button>
        </Box>
      )}
      {isUploadDenyAvailable && (
        <Box mt={2} mr={6}>
          <DenyUploadRequirement />
        </Box>
      )}
    </Box>
  );
}
