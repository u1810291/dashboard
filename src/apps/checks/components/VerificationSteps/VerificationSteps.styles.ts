import { makeStyles } from '@material-ui/core/styles';

export const useStyles = makeStyles(() => ({
  label: {
    fontSize: 16,
  },
  value: {
    padding: 8,
    border: '1px solid var(--mgi-theme-palette-lightgray)',
    '& + &': {
      marginTop: -1,
    },
    '&:first-child': {
      borderTopLeftRadius: 4,
      borderTopRightRadius: 4,
    },
    '&:last-child': {
      borderBottomLeftRadius: 4,
      borderBottomRightRadius: 4,
    },
  },
}));
