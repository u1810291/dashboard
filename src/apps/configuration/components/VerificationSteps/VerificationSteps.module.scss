.marginBlock {
  margin-top: 15px;
}

.marginLeft {
  margin-left: 14px;
}

.docWrapperMany, .docWrapper {
  .child {
    width: 305px;
    height: 49px;
    border: 1px solid var(--mgi-theme-palette-lightgray);
    box-sizing: border-box;
    border-radius: 4px;

    span {
      padding-left: 10px;
      font-size: 16px;
      line-height: 49px;
      color: var(--mgi-theme-palette-darkergray);
    }
  }
}

.docWrapperMany {
  .child {
    border-radius: 0;
    border-bottom: none;
  }

  > div:first-child {
    .child {
      border-radius: 4px 4px 0 0;
    }
  }

  > div:last-child {
     .child {
       border: 1px solid var(--mgi-theme-palette-lightgray);
       border-radius:  0 0 4px 4px;
     }
   }
}

.newStep {
  border: none;
  color: var(--mgi-theme-palette-blue);
  text-decoration: underline;
  margin-left: 0;
  padding-left: 0;
}

.docTitle {
  margin-bottom: 5px;
  margin-left: 0;
  color: var(--mgi-theme-palette-darkergray);
}

.verificationSteps {
  .verificationInfo {
    margin-top: 20px;

    > div:first-child {
      margin-top: 40px;
    }
  }
}
