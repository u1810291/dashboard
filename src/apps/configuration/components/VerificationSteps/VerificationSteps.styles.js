import { makeStyles } from '@material-ui/core/styles';

export const useStyles = makeStyles(() => ({
  verificationSteps: {
    '& .docWrapperMany, & .docWrapper': {
      '& .child': {
        width: 305,
        height: 49,
        border: [[1, 'solid', 'var(--mgi-theme-palette-lightgray)']],
        boxSizing: 'border-box',
        borderRadius: 4,
        '& span': {
          paddingLeft: 10,
          fontSize: 16,
          lineHeight: '49px',
          color: 'var(--mgi-theme-palette-darkergray)',
        },
      },
    },
    '& .docWrapperMany': {
      '& .child': {
        borderRadius: 0,
        borderBottom: 'none',
      },
      '& > div:first-child': {
        '& .child': {
          borderRadius: [[4, 4, 0, 0]],
        },
      },
      '& > div:last-child': {
        '& .child': {
          border: [[1, 'solid', 'var(--mgi-theme-palette-lightgray)']],
          borderRadius: [[0, 0, 4, 4]],
        },
      },
    },
  },
  verificationInfo: {
    marginTop: 20,
    '& > div:first-child': {
      marginTop: 40,
    },
  },
  docTitle: {
    marginBottom: 5,
    marginLeft: 0,
  },
  newStep: {
    border: 'none',
    color: 'var(--mgi-theme-palette-blue)',
    textDecoration: 'underline',
    marginLeft: 0,
    paddingLeft: 0,
  },
}));
