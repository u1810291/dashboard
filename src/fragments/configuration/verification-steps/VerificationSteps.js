import classNames from 'classnames';
import { Button, Items, Text } from 'components';
import confirm from 'components/confirm';
import { closeOverlay, createOverlay } from 'components/overlay';
import { difference, without } from 'lodash';
import React, { useState } from 'react';
import { FiTrash2, FiEdit2 } from 'react-icons/fi';
import { useIntl } from 'react-intl';
import { AVAILABLE_DOCUMENT_TYPES } from 'state/merchant/merchant.model';
import VerificationStepModal from '../verification-steps-modal';
import CSS from './VerificationSteps.module.scss';

export function removeItem(steps, index) {
  const updatedSteps = [...steps];
  updatedSteps.splice(index, 1);
  return updatedSteps;
}

export function replaceItem(steps, index, values) {
  const updatedSteps = [...steps];
  updatedSteps[index] = values;
  return updatedSteps;
}

export function addItem(steps, values) {
  const updatedSteps = [...steps];
  updatedSteps.push(values);
  return updatedSteps;
}

export function accessibleItems(available, mandatory, steps, index) {
  return difference(available, mandatory, ...without(steps, steps[index]));
}

export function VerificationSteps({
  steps = [],
  mandatoryDocumentTypes = [],
  onChange = () => {},
}) {
  const [availableDocumentTypes] = useState(AVAILABLE_DOCUMENT_TYPES);
  const intl = useIntl();

  const onRemoveItem = async (index) => {
    try {
      await confirm(intl.formatMessage({ id: 'confirm_string' }));
      onChange({ verificationSteps: removeItem(steps, index) });
    } catch (e) {
      // none, canceled
    }
  };

  const onEditItem = (index) => {
    createOverlay(
      <VerificationStepModal
        values={index !== undefined ? steps[index] : []}
        items={accessibleItems(
          availableDocumentTypes,
          mandatoryDocumentTypes,
          steps,
          index,
        )}
        onSave={(values) => {
          closeOverlay();
          onChange({
            verificationSteps:
              index !== undefined
                ? replaceItem(steps, index, values)
                : addItem(steps, values),
          });
        }}
      />,
    );
  };

  return (
    <fieldset className="mgi-fieldset">
      <div className={CSS.verificationSteps}>
        <Text size={3} weight={4}>
          {intl.formatMessage({ id: 'flow.documentTypeStep.title' })}
        </Text>
        {mandatoryDocumentTypes.map((doc, index) => (
          // eslint-disable-next-line react/no-array-index-key
          <fieldset className="mgi-fieldset" key={index}>
            <legend className="text-active">
              {intl.formatMessage({ id: 'flow.documentTypeStep.stepNo' })}
              {' '}
              {index + 1}
            </legend>
            <Items templateColumns="minmax(auto, 100%) auto" justifyContent="start">
              {intl.formatMessage({ id: `flow.documentTypeStep.${doc}` })}
              <span className="text-secondary">
                {intl.formatMessage({ id: 'required' })}
              </span>
            </Items>
          </fieldset>
        ))}
        {steps.map((step, index) => (
          <div
            key={index} // eslint-disable-line react/no-array-index-key
            className={CSS.verificationInfo}
          >
            <div className={classNames('text-active', [CSS.docTitle])}>
              <Items templateColumns="minmax(auto, 100%) auto">
                <Text size={2} weight={4}>
                  {intl.formatMessage({ id: 'flow.documentTypeStep.stepNo' })}
                  {' - '}
                  {index + mandatoryDocumentTypes.length + 1}
                </Text>
              </Items>
            </div>
            <Items gap={1} align="center">
              <Items
                flow="row"
                gap={step.length > 1 ? 0 : 1}
                className={classNames({
                  [CSS.docWrapper]: step.length === 1,
                  [CSS.docWrapperMany]: step.length > 1,
                })}
              >
                {step.sort().map((doc) => (
                  <Items flow="row" key={doc} className={[CSS.children]}>
                    <div className={[CSS.child]}>
                      <Text>
                        {intl.formatMessage({ id: `flow.documentTypeStep.${doc}` })}
                      </Text>
                    </div>
                  </Items>
                ))}
              </Items>
              <Items flow="row" gap={1}>
                <Items flow="column">
                  <Button buttonStyle="invisible">
                    <FiEdit2 color="color-primary" onClick={() => onEditItem(index)} />
                  </Button>
                  <Button buttonStyle="invisible" data-role="deleteVerificationStep">
                    <FiTrash2 size="1rem" className="color-red" onClick={() => onRemoveItem(index)} />
                  </Button>
                </Items>
              </Items>
            </Items>
          </div>
        ))}
      </div>
      {difference(availableDocumentTypes, mandatoryDocumentTypes, ...steps).length > 0 && (
        <Button
          className={CSS.newStep}
          onClick={() => onEditItem()}
          disabled={difference(availableDocumentTypes, mandatoryDocumentTypes, ...steps).length === 0}
          data-role="newVerificationStep"
        >
          {intl.formatMessage({ id: 'flow.documentTypeStep.button.title' })}
        </Button>
      )}
    </fieldset>
  );
}
