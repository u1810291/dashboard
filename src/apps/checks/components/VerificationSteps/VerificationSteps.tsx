import { Box, Button, IconButton, Typography } from '@material-ui/core';
import { useConfirm, useOverlay } from 'apps/overlay';
import { difference, without } from 'lodash';
import React, { useCallback, useState, useMemo, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { FiEdit2, FiTrash2 } from 'react-icons/fi';
import { LoadableAdapter } from 'lib/Loadable.adapter';
import { useIntl } from 'react-intl';
import { CustomDocumentResponse } from 'models/CustomDocument.model';
import { DocumentListOrdered, DocumentTypes, isCustomDocument } from 'models/Document.model';
import { IFlow } from 'models/Flow.model';
import { MerchantTags } from 'models/Merchant.model';
import { merchantCustomDocumentsLoad } from 'state/merchant/merchant.actions';
import { selectMerchantCustomDocumentsModel, selectMerchantModel, selectMerchantTags } from 'state/merchant/merchant.selectors';
import { VerificationStepsModal } from '../VerificationStepsModal/VerificationStepsModal';
import { useStyles } from './VerificationSteps.styles';
import { DenyUploadRequirement } from '../../../configuration/containers/DenyUploadRequirement/DenyUploadRequirement';

// TODO: use sync with flowbuilder types when product is ready
type CustomizableDocumentType = DocumentTypes | string;
type CustomizableVerificationSteps = CustomizableDocumentType[][];

export function removeItem(steps: CustomizableVerificationSteps, index: number): CustomizableVerificationSteps {
  const updatedSteps = [...steps];
  updatedSteps.splice(index, 1);
  return updatedSteps;
}

export function replaceItem(steps: CustomizableVerificationSteps, index: number, values: CustomizableDocumentType[]): CustomizableVerificationSteps {
  const updatedSteps = [...steps];
  updatedSteps[index] = values;
  return updatedSteps;
}

export function addItem(steps: CustomizableVerificationSteps, values: CustomizableDocumentType[]) {
  const updatedSteps = [...steps];
  updatedSteps.push(values);
  return updatedSteps;
}

export function addItemBefore(steps: CustomizableVerificationSteps, index: number, values: CustomizableDocumentType[]): CustomizableVerificationSteps {
  const updatedSteps = [...steps];
  updatedSteps.splice(index, 0, values);
  return updatedSteps;
}

export function accessibleItems(available: DocumentTypes[], steps: CustomizableVerificationSteps, index: number): string[] {
  return difference(available, ...without(steps, steps[index]));
}

export function accessibleCustomItems(available: CustomDocumentResponse[], selected: string[][]): CustomDocumentResponse[] {
  const alreadyUsed = selected.flatMap((item) => item);
  return available.filter(({ type }: CustomDocumentResponse) => !alreadyUsed.includes(type));
}

export function VerificationSteps({
  steps = [],
  onChange,
}: {
  steps: CustomizableVerificationSteps;
  onChange: ({ verificationSteps }: IFlow) => void;
}) {
  const intl = useIntl();
  const classes = useStyles();
  const dispatch = useDispatch();
  const merchantModel = useSelector(selectMerchantModel);
  const merchantCustomDocumentsModel = useSelector(selectMerchantCustomDocumentsModel);
  const tags: MerchantTags[] = useSelector(selectMerchantTags);
  const [createOverlay, closeOverlay] = useOverlay();
  const confirm = useConfirm(intl.formatMessage({ id: 'confirm_string' }));
  const [availableDocumentTypes] = useState<DocumentTypes[]>(DocumentListOrdered);

  const isCustomDocumentAvailable = useMemo(() => tags.includes(MerchantTags.CanUseCustomDocument), [tags]);

  const { regular, custom } = useMemo(() => steps.reduce((result, step) => {
    if (step.some((documentType) => isCustomDocument(documentType))) {
      result.custom.push(step);
    } else {
      result.regular.push(step);
    }

    return result;
  }, { regular: [], custom: [] }), [steps]);

  const customDocumentsMap = useMemo((): { [key: string]: string } => {
    if (!merchantCustomDocumentsModel?.value) {
      return {};
    }

    return merchantCustomDocumentsModel.value.reduce((result: { [key: string]: string }, customDocumentType: CustomDocumentResponse) => ({
      ...result,
      [customDocumentType.type]: customDocumentType.name,
    }), {});
  }, [merchantCustomDocumentsModel]);

  const avalibleCustomDocuments = useMemo(() => accessibleCustomItems(merchantCustomDocumentsModel?.value, custom), [merchantCustomDocumentsModel, custom]);

  const handleRemoveItem = useCallback(
    async (index) => {
      try {
        await confirm();
        onChange({ verificationSteps: removeItem(steps, index) });
      } catch (e) {
        // none, canceled
      }
    },
    [confirm, onChange, steps],
  );

  const onEditRegularDocument = useCallback(
    (index?: number) => () => {
      createOverlay(
        <VerificationStepsModal
          mode="regular"
          values={index !== undefined ? regular[index] : []}
          documents={accessibleItems(availableDocumentTypes, steps, index)}
          onSave={(values: string[]) => {
            closeOverlay();
            onChange({
              verificationSteps: index !== undefined ? replaceItem(steps, index, values) : addItemBefore(steps, regular.length, values),
            });
          }}
        />,
      );
    },
    [createOverlay, onChange, availableDocumentTypes, closeOverlay, steps, regular],
  );

  const onEditCustomDocument = useCallback(
    (index?: number) => () => {
      createOverlay(
        <VerificationStepsModal
          mode="custom"
          values={index !== undefined ? custom[index] : []}
          customDocuments={avalibleCustomDocuments}
          onSave={(values: string[]) => {
            closeOverlay();
            onChange({
              verificationSteps: index !== undefined ? replaceItem(steps, regular.length + index, values) : addItem(steps, values),
            });
          }}
        />,
      );
    },
    [createOverlay, custom, avalibleCustomDocuments, closeOverlay, onChange, steps, regular.length],
  );

  useEffect(() => {
    if (merchantModel.isLoaded && LoadableAdapter.isPristine(merchantCustomDocumentsModel) && !merchantCustomDocumentsModel.value?.length) {
      try {
        dispatch(merchantCustomDocumentsLoad());
      } catch (error) {
        console.error(error);
      }
    }
  }, [dispatch, merchantCustomDocumentsModel, merchantModel]);

  return (
    <Box>
      <Typography variant="h4" gutterBottom>
        {intl.formatMessage({ id: 'flow.documentTypeStep.title' })}
      </Typography>

      {/* regular documents */}
      {regular.map((step, index) => (
        <React.Fragment key={index}>
          <Box mt={2}>
            <Typography variant="h6" color="primary" gutterBottom>
              {intl.formatMessage({ id: 'flow.documentTypeStep.stepNo' })}
              {' - '}
              {index + 1}
            </Typography>
            <Box display="flex" mt={1}>
              {/* step list */}
              <Box flexGrow={1}>
                {step.sort().map((item: DocumentTypes) => (
                  <Box key={item} className={classes.value}>
                    <Typography variant="body1" gutterBottom>
                      {intl.formatMessage({ id: `flow.documentTypeStep.${item}` })}
                    </Typography>
                  </Box>
                ))}
              </Box>
              {/* actions */}
              <Box flexGrow={0} display="flex" ml={1} alignItems="center">
                <IconButton size="small" color="primary" onClick={onEditRegularDocument(index)}>
                  <FiEdit2 />
                </IconButton>
                <Box ml={1}>
                  <IconButton size="small" onClick={() => handleRemoveItem(index)}>
                    <FiTrash2 className="color-red" />
                  </IconButton>
                </Box>
              </Box>
            </Box>
          </Box>
        </React.Fragment>
      ))}

      {difference(availableDocumentTypes, ...steps).length > 0 && (
        <Box mt={2}>
          <Button
            variant="text"
            color="primary"
            onClick={onEditRegularDocument()}
            disabled={difference(availableDocumentTypes, ...steps).length === 0}
          >
            {intl.formatMessage({ id: 'flow.documentTypeStep.button.title' })}
          </Button>
        </Box>
      )}

      {/* custom documents */}
      {(isCustomDocumentAvailable || (custom.length > 0)) ? (
        <>
          <Typography variant="h6" color="primary">
            {intl.formatMessage({ id: 'CustomDocuments.settings.customDocuments' })}
          </Typography>

          {custom.map((step, index) => (
            <React.Fragment key={index}>
              <Box mt={2}>
                <Typography variant="h6" color="primary" gutterBottom>
                  {intl.formatMessage({ id: 'flow.documentTypeStep.stepNo' })}
                  {' - '}
                  {index + 1}
                </Typography>

                <Box display="flex" mt={1}>
                  {/* step list */}
                  <Box flexGrow={1}>
                    {step.sort().map((item: string) => (
                      <Box key={item} className={classes.value}>
                        <Typography variant="body1" gutterBottom>
                          {customDocumentsMap[item]}
                        </Typography>
                      </Box>
                    ))}
                  </Box>
                  {/* actions */}
                  <Box flexGrow={0} display="flex" ml={1} alignItems="center">
                    {isCustomDocumentAvailable && (
                      <IconButton size="small" color="primary" onClick={onEditCustomDocument(index)}>
                        <FiEdit2 />
                      </IconButton>
                    )}
                    <Box ml={1}>
                      <IconButton size="small" onClick={() => handleRemoveItem(index)}>
                        <FiTrash2 className="color-red" />
                      </IconButton>
                    </Box>
                  </Box>
                </Box>
              </Box>
            </React.Fragment>
          ))}

          {isCustomDocumentAvailable && (avalibleCustomDocuments.length > 0) && (
            <Box mt={2}>
              <Button variant="text" color="primary" onClick={onEditCustomDocument()}>
                {intl.formatMessage({ id: 'CustomDocuments.settings.addVerificationRequirement' })}
              </Button>
            </Box>
          )}
        </>
      ) : null}

      <Box mt={2} mr={6}>
        <DenyUploadRequirement />
      </Box>
    </Box>
  );
}
