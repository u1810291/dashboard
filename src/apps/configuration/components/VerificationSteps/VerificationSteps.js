import clsx from 'clsx';
import { useIntl } from 'react-intl';
import React, { useState } from 'react';
import { difference, without } from 'lodash';
import { IconButton } from '@material-ui/core';
import { Button, Items, Text } from 'components';
import confirm from 'components/confirm';
import { closeOverlay, createOverlay } from 'components/overlay';
import { FiEdit2, FiTrash2 } from 'react-icons/fi';
import { AVAILABLE_DOCUMENT_TYPES } from 'state/merchant/merchant.model';
import VerificationStepModal from '../VerificationStepsModal';
import { useStyles } from './VerificationSteps.styles';

export function removeItem(steps, index) {
  const updatedSteps = [...steps];
  updatedSteps.splice(index, 1);
  return updatedSteps;
}

export function replaceItem(steps, index, values) {
  const updatedSteps = [...steps];
  updatedSteps[index] = values;
  return updatedSteps;
}

export function addItem(steps, values) {
  const updatedSteps = [...steps];
  updatedSteps.push(values);
  return updatedSteps;
}

export function accessibleItems(available, mandatory, steps, index) {
  return difference(available, mandatory, ...without(steps, steps[index]));
}

export function VerificationSteps({
  steps = [],
  mandatoryDocumentTypes = [],
  onChange = () => {},
}) {
  const intl = useIntl();
  const classes = useStyles();
  const [availableDocumentTypes] = useState(AVAILABLE_DOCUMENT_TYPES);

  const onRemoveItem = async (index) => {
    try {
      await confirm(intl.formatMessage({ id: 'confirm_string' }));
      onChange({ verificationSteps: removeItem(steps, index) });
    } catch (e) {
      // none, canceled
    }
  };

  const onEditItem = (index) => {
    createOverlay(
      <VerificationStepModal
        values={index !== undefined ? steps[index] : []}
        items={accessibleItems(
          availableDocumentTypes,
          mandatoryDocumentTypes,
          steps,
          index,
        )}
        onSave={(values) => {
          closeOverlay();
          onChange({
            verificationSteps:
              index !== undefined
                ? replaceItem(steps, index, values)
                : addItem(steps, values),
          });
        }}
      />,
    );
  };

  return (
    <fieldset className="mgi-fieldset">
      <div className={classes.verificationSteps}>
        <Text size={3} weight={4}>
          {intl.formatMessage({ id: 'flow.documentTypeStep.title' })}
        </Text>
        {mandatoryDocumentTypes.map((doc, index) => (
          // eslint-disable-next-line react/no-array-index-key
          <fieldset className="mgi-fieldset" key={index}>
            <legend className="text-active">
              {intl.formatMessage({ id: 'flow.documentTypeStep.stepNo' })}
              {' '}
              {index + 1}
            </legend>
            <Items templateColumns="minmax(auto, 100%) auto" justifyContent="start">
              {intl.formatMessage({ id: `flow.documentTypeStep.${doc}` })}
              <span className="text-secondary">
                {intl.formatMessage({ id: 'required' })}
              </span>
            </Items>
          </fieldset>
        ))}
        {steps.map((step, index) => (
          <div
            key={index} // eslint-disable-line react/no-array-index-key
            className={classes.verificationInfo}
          >
            <div className={clsx('text-active', [classes.docTitle])}>
              <Items templateColumns="minmax(auto, 100%) auto">
                <Text size={2} weight={4}>
                  {intl.formatMessage({ id: 'flow.documentTypeStep.stepNo' })}
                  {' - '}
                  {index + mandatoryDocumentTypes.length + 1}
                </Text>
              </Items>
            </div>
            <Items gap={1} align="center">
              <Items
                flow="row"
                gap={step.length > 1 ? 0 : 1}
                className={clsx({
                  docWrapper: step.length === 1,
                  docWrapperMany: step.length > 1,
                })}
              >
                {step.sort().map((doc) => (
                  <Items flow="row" key={doc} className="children">
                    <div className="child">
                      <Text>
                        {intl.formatMessage({ id: `flow.documentTypeStep.${doc}` })}
                      </Text>
                    </div>
                  </Items>
                ))}
              </Items>
              <Items flow="row" gap={1}>
                <Items flow="column">
                  <IconButton size="small" color="primary" onClick={() => onEditItem(index)}>
                    <FiEdit2 />
                  </IconButton>
                  <IconButton size="small" onClick={() => onRemoveItem(index)}>
                    <FiTrash2 className="color-red" />
                  </IconButton>
                </Items>
              </Items>
            </Items>
          </div>
        ))}
      </div>
      {difference(availableDocumentTypes, mandatoryDocumentTypes, ...steps).length > 0 && (
        <Button
          className={classes.newStep}
          onClick={() => onEditItem()}
          disabled={difference(availableDocumentTypes, mandatoryDocumentTypes, ...steps).length === 0}
          data-role="newVerificationStep"
        >
          {intl.formatMessage({ id: 'flow.documentTypeStep.button.title' })}
        </Button>
      )}
    </fieldset>
  );
}
