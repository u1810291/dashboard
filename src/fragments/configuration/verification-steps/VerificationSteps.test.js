import React from 'react'
import { shallow } from 'enzyme'
import VerificationSteps, {
  removeItem,
  replaceItem,
  addItem,
  accessibleItems
} from './VerificationSteps'

describe('VerificationSteps', () => {
  const availableDocumentTypes = [
    'passport',
    'national-id',
    'driving-license',
    'proof-of-residency'
  ]

  const mandatoryDocumentTypes = ['liveness']

  test('mandatory steps', () => {
    const component = shallow(
      <VerificationSteps
        availableDocumentTypes={availableDocumentTypes}
        mandatoryDocumentTypes={mandatoryDocumentTypes}
      />
    )

    expect(component).toMatchSnapshot()
  })

  test('steps ', () => {
    const component = shallow(
      <VerificationSteps
        availableDocumentTypes={availableDocumentTypes}
        steps={[['national-id', 'driving-license'], ['proof-of-residency']]}
      />
    )

    expect(component).toMatchSnapshot()
  })
})

describe('helpers', () => {
  test('removeItem', () => {
    expect(removeItem([1, 2, 3, 4], 2)).toEqual([1, 2, 4])
  })

  test('replaceItem', () => {
    expect(replaceItem([1, 2, 3, 4], 2, 'new')).toEqual([1, 2, 'new', 4])
  })

  test('addItem', () => {
    expect(addItem([1, 2, 3, 4], 'new')).toEqual([1, 2, 3, 4, 'new'])
  })

  test('accessibleItems', () => {
    const mandatory = ['a']
    const available = ['b', 'c', 'd', 'e', 'f']
    const steps = [['b', 'c'], ['d']]
    expect(accessibleItems(available, mandatory, steps, 0)).toEqual([
      'b',
      'c',
      'e',
      'f'
    ])
    expect(accessibleItems(available, mandatory, steps, 1)).toEqual([
      'd',
      'e',
      'f'
    ])
    expect(accessibleItems(available, mandatory, steps)).toEqual(['e', 'f'])
  })
})
