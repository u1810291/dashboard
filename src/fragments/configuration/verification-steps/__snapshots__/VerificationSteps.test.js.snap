// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerificationSteps mandatory steps 1`] = `
<fieldset
  className="mgi-fieldset"
>
  <legend>
    <h3>
      <FormattedMessage
        id="flow.documentTypeStep.title"
        values={Object {}}
      />
    </h3>
  </legend>
  <fieldset
    className="mgi-fieldset"
    key="0"
  >
    <legend
      className="text-active"
    >
      <FormattedMessage
        id="flow.documentTypeStep.stepNo"
        values={Object {}}
      />
       
      1
    </legend>
    <div
      className="mgi-items"
    >
      <span
        className="mgi-items--grow"
      >
        <FormattedMessage
          id="flow.documentTypeStep.liveness"
          values={Object {}}
        />
      </span>
      <span
        className="text-secondary"
      >
        <FormattedMessage
          id="required"
          values={Object {}}
        />
      </span>
    </div>
  </fieldset>
  <Button
    buttonStyle="primary"
    disabled={false}
    onClick={[Function]}
  >
    Add another step
  </Button>
</fieldset>
`;

exports[`VerificationSteps steps  1`] = `
<fieldset
  className="mgi-fieldset"
>
  <legend>
    <h3>
      <FormattedMessage
        id="flow.documentTypeStep.title"
        values={Object {}}
      />
    </h3>
  </legend>
  <fieldset
    className="mgi-fieldset"
    key="0"
  >
    <legend
      className="text-active"
    >
      <div
        className="mgi-items"
      >
        <span
          className="mgi-items--grow"
        >
          <FormattedMessage
            id="flow.documentTypeStep.stepNo"
            values={Object {}}
          />
           
          1
        </span>
        <div
          className="mgi-items mgi-items--narrow"
        >
          <Button
            buttonStyle="invisible"
          >
            <pencil.svg
              className="svg-active"
              onClick={[Function]}
            />
          </Button>
          <Button
            buttonStyle="invisible"
          >
            <trashbin.svg
              className="svg-error"
              onClick={[Function]}
            />
          </Button>
        </div>
      </div>
    </legend>
    <div>
      <FormattedMessage
        id="flow.documentTypeStep.driving-license"
        values={Object {}}
      />
    </div>
    <div
      className="text-secondary"
    >
      <FormattedMessage
        id="or"
        values={Object {}}
      />
    </div>
    <div>
      <FormattedMessage
        id="flow.documentTypeStep.national-id"
        values={Object {}}
      />
    </div>
  </fieldset>
  <fieldset
    className="mgi-fieldset"
    key="1"
  >
    <legend
      className="text-active"
    >
      <div
        className="mgi-items"
      >
        <span
          className="mgi-items--grow"
        >
          <FormattedMessage
            id="flow.documentTypeStep.stepNo"
            values={Object {}}
          />
           
          2
        </span>
        <div
          className="mgi-items mgi-items--narrow"
        >
          <Button
            buttonStyle="invisible"
          >
            <pencil.svg
              className="svg-active"
              onClick={[Function]}
            />
          </Button>
          <Button
            buttonStyle="invisible"
          >
            <trashbin.svg
              className="svg-error"
              onClick={[Function]}
            />
          </Button>
        </div>
      </div>
    </legend>
    <div>
      <FormattedMessage
        id="flow.documentTypeStep.proof-of-residency"
        values={Object {}}
      />
    </div>
  </fieldset>
  <Button
    buttonStyle="primary"
    disabled={false}
    onClick={[Function]}
  >
    Add another step
  </Button>
</fieldset>
`;
